java/src/main/native/src/RmmJni.cpp

```diff
@@ -109,11 +110,15 @@ class tracking_resource_adaptor final : public base_tracking_resource_adaptor {
     auto result = resource->allocate(num_bytes, stream);
     if (result) {
       total_allocated += num_bytes;
-      local_allocated += num_bytes;
 
-      // Note: this is not thread safe.
-      max_outstanding.store(std::max(total_allocated, max_outstanding));
-      local_max_outstanding.store(std::max(local_allocated, local_max_outstanding));
+      // Note: none of the below is thread safe. It is only meaningful when
+      // a single thread is used.
+      max_outstanding = std::max(total_allocated.load(), max_outstanding);
+
+      // `total_allocated - local_allocated` can be negative in the case where we free
+      // after we call `reset_local_max_outstanding`
+      std::size_t local_diff = std::max(static_cast<long>(total_allocated - local_allocated), 0L);
```

Comment:
Maybe use `static_cast<intptr_t>` instead of `static_cast<long>`
I don't think long is guaranteed to be big enough to hold a size_t.  You could use `long long` as well.

Reply:
So I think `long` == `long long` (I think this is 32bit vs 64bit compiled programs). To cover all of `std::size_t`, I'd have to go to `unsigned long`. That's a lot of GPU memory ;) I am not sure we need to worry too much about that, especially since we are going to send this to Spark shortly, which runs java, and java's long is 64-bit and signed.

```
size_t max value: 18446744073709551615
long max value: 9223372036854775807
unsigned long max value: 18446744073709551615
long long max value: 9223372036854775807
unsigned long long max value: 18446744073709551615
```

Reply:
I think in this case, long is sufficient because we are an LP64 architecture (we don't run on Windows, do we?).
`std::intptr_t` is guaranteed to be the same width as `std::size_t`, but signed (I don't think `ssize_t` is standard?).  You could use int64_t here, since as you say we know we are going to pass it to java, which is using 64 bits.  This was more of a technical nit, than an actual concern that it will break (too much history cross-porting to different architectures...)

Reply:
These types are very confusing which can be alias of other types depending on the system. Therefore, for clarity, please use the fix-width types: `(u)int32_t` and `(u)int64_t`. They guarantee you to have known limits.

Reply:
So this is a debug feature, we don't use `(u)int*` right now (e.g. I'd make things more inconsistent unless I change the whole thing), and I am not sure whether cuDF is moving away from the alias types. I think we can update in one PR that is "go away from these old types to the better ones" in the future.